checks.d/splunk_topology.py

```diff
@@ -0,0 +1,190 @@
+        self.timeout = float(instance.get('timeout', default_timeout))
+        self.max_retry_count = int(instance.get('max_retry_count', max_retry_count))
+        self.seconds_between_retries = int(instance.get('seconds_between_retries', seconds_between_retries))
+        self.polling_interval = int(instance.get('polling_interval', polling_interval))
+
+    def get_auth_tuple(self):
+        return self.username, self.password
+
+
+class Instance:
+    INSTANCE_TYPE = "splunk"
+
+    def __init__(self, instance, init_config):
+        self.instance_config = InstanceConfig(instance, init_config)
+        self.saved_searches = [SavedSearch(saved_search) for saved_search in instance['saved_searches']]
+        self.instance_key = {
+            "type": self.INSTANCE_TYPE,
+            "url": self.instance_config.base_url
+        }
+        self.tags = instance.get('tags', [])
+        self.last_successful_poll = 0
+
+    def should_poll(self, time_seconds):
+        return self.last_successful_poll == 0 or time_seconds >= self.last_successful_poll + self.instance_config.polling_interval
+
+
+class SplunkTopology(AgentCheck):
+    # SERVICE_CHECK_NAME = "splunk.topology_information"
+    # service_check_needed = True
+
+    def __init__(self, name, init_config, agentConfig, instances=None):
+        super(SplunkTopology, self).__init__(name, init_config, agentConfig, instances)
+        # Data to keep over check runs, keyed by instance url
+        self.instance_data = dict()
+
+    def check(self, instance):
+        if 'url' not in instance:
+            raise CheckException('Splunk topology instance missing "url" value.')
+
+        if instance["url"] not in self.instance_data:
+            self.instance_data[instance["url"]] = Instance(instance, self.init_config)
+
+        instance = self.instance_data[instance["url"]]
+        current_time = self._current_time_seconds()
+        if not instance.should_poll(current_time):
+            return
+
+        instance_key = instance.instance_key
+
+        search_ids = [(self._dispatch_saved_search(instance.instance_config, saved_search), saved_search)
+                      for saved_search in instance.saved_searches]
+
+        self.start_snapshot(instance_key)
+
+        try:
+            for (sid, saved_search) in search_ids:
+                response = self._search(instance.instance_config, sid)
+                if saved_search.element_type == "component":
+                    self._extract_components(instance, response)
+                elif saved_search.element_type == "relation":
+                    self._extract_relations(instance, response)
+
+            # If everything was successful, update the timestamp
+            instance.last_successful_poll = current_time
+        finally:
+            self.stop_snapshot(instance_key)
+
+    def _current_time_seconds(self):
+        return int(round(time.time() * 1000))
+
+    def _search(self, instance_config, search_id):
```

Comment:
The search function, currently keeps doing request until a search job is complete. When there is one job that is very slow, and some that are very fast, it might happen that we block on the slow job if that happens to be first. The results of the fast jobs then are collected later, while they could allready have been collected. This is not neccesarily wrong but not the most performant way of doing it. Apparently saved searched also have a ttl, which is by default 10 minutes (see https://www.splunk.com/blog/2012/09/12/how-long-does-my-search-live-default-search-ttl/).